src/index.ts

```diff
@@ -0,0 +1,134 @@
+  ),
+});
+
+const createReducer = (definition: Machine.Definition.Impl) => {
+  let log = createLogger(definition);
+  return (machineState: Machine.State.Impl, internalEvent: InternalEvent): Machine.State.Impl => {
+    if (internalEvent.type === "SET_CONTEXT") {
+      let nextContext = internalEvent.updater(machineState.context);
+      log("Context update", ["Previous Context", machineState.context], ["Next Context", nextContext]);
+
+      return { ...machineState, context: nextContext };
+    }
+
+    if (internalEvent.type === "SEND") {
+      let sendable = internalEvent.sendable;
+      let event = typeof sendable === "string" ? { type: sendable } : sendable;
+      let context = machineState.context;
+      let stateNode = R.get(definition.states, machineState.value)!;
+      let resolvedTransition =
+        R.get(R.fromMaybe(stateNode.on), event.type) ?? R.get(R.fromMaybe(definition.on), event.type);
+
+      if (!resolvedTransition) {
+        log(
+          `Current state doesn't listen to event type "${event.type}".`,
+          ["Current State", machineState],
+          ["Event", event]
+        );
+        return machineState;
+      }
+
+      let [nextStateValue, didGuardDeny = false] = (() => {
+        if (typeof resolvedTransition === "string") return [resolvedTransition];
+        if (resolvedTransition.guard === undefined) return [resolvedTransition.target];
+        if (resolvedTransition.guard({ context, event })) return [resolvedTransition.target];
+        return [resolvedTransition.target, true]
+      })() as [Machine.StateValue.Impl, true?]
+
+      if (didGuardDeny) {
+        log(
+          `Transition from "${machineState.value}" to "${nextStateValue}" denied by guard`,
+          ["Event", event],
+          ["Context", context]
+        );
+        return machineState;
+      }
+      log(`Transition from "${machineState.value}" to "${nextStateValue}"`, ["Event", event]);
+
+      return {
+        value: nextStateValue,
+        context,
+        event,
+        nextEvents: R.keys(R.concat(R.fromMaybe(stateNode.on), R.fromMaybe(definition.on))),
+      };
+    }
+
+    return assertNever(internalEvent);
+  };
+};
+
+interface SetContextEvent {
+  type: "SET_CONTEXT";
+  updater: Machine.ContextUpdater.Impl;
+}
+
+interface SendEvent {
+  type: "SEND";
+  sendable: Machine.Sendable.Impl;
+}
+
+type InternalEvent = SetContextEvent | SendEvent;
+
+export type Console =
+  { log: (a: string, b: string | object) => void
+  , groupCollapsed?: (...l: string[]) => void
+  , groupEnd?: () => void
+  }
+
+// test sub-typing
+const defaultConsole: Console = console
+
+const createLogger = (definition: Machine.Definition.Impl) => (groupLabel: string, ...nested: [string, string | object][]) => {
+  if (!definition.verbose) return;
+  
+  let console = definition.console || defaultConsole
+  if (process.env.NODE_ENV === "development" || process.env.NODE_ENV === "test") {
```

Comment:
Can we remove this? Not sure what's the point, we already have `verbose`. If someone passed `verbose: true` for prod, it's their fault :P

```suggestion
```

Reply:
Oh, this signals babel that it's safe to tree shake this bit of code. We get a few extra bites smalles.

Reply:
Ah nice in that case I should wrap the whole body of the function in `if`?